pkg/auth/dependency/interaction/flows/webapp.go

```diff
@@ -230,6 +258,111 @@ func (f *WebAppFlow) TriggerOOBOTP(token string, step interaction.Step) (*WebApp
 	}, nil
 }
 
+func (f *WebAppFlow) AddLoginID(userID string, loginID loginid.LoginID) (result *WebAppResult, err error) {
+	clientID := ""
+	i, err := f.Interactions.NewInteractionAddIdentity(&interaction.IntentAddIdentity{
+		Identity: identity.Spec{
+			Type: authn.IdentityTypeLoginID,
+			Claims: map[string]interface{}{
+				identity.IdentityClaimLoginIDKey:   loginID.Key,
+				identity.IdentityClaimLoginIDValue: loginID.Value,
+			},
+		},
+	}, clientID, userID)
+	if err != nil {
+		return
+	}
+
+	return f.afterAddUpdateRemoveLoginID(i)
+}
+
+func (f *WebAppFlow) RemoveLoginID(userID string, loginID loginid.LoginID) (result *WebAppResult, err error) {
+	clientID := ""
+	i, err := f.Interactions.NewInteractionRemoveIdentity(&interaction.IntentRemoveIdentity{
+		Identity: identity.Spec{
+			Type: authn.IdentityTypeLoginID,
+			Claims: map[string]interface{}{
+				identity.IdentityClaimLoginIDKey:   loginID.Key,
+				identity.IdentityClaimLoginIDValue: loginID.Value,
+			},
+		},
+	}, clientID, userID)
+	if err != nil {
+		return
+	}
+
+	return f.afterAddUpdateRemoveLoginID(i)
+}
+
+func (f *WebAppFlow) UpdateLoginID(userID string, oldLoginID loginid.LoginID, newLoginID loginid.LoginID) (result *WebAppResult, err error) {
+	clientID := ""
+	i, err := f.Interactions.NewInteractionUpdateIdentity(&interaction.IntentUpdateIdentity{
+		OldIdentity: identity.Spec{
+			Type: authn.IdentityTypeLoginID,
+			Claims: map[string]interface{}{
+				identity.IdentityClaimLoginIDKey:   oldLoginID.Key,
+				identity.IdentityClaimLoginIDValue: oldLoginID.Value,
+			},
+		},
+		NewIdentity: identity.Spec{
+			Type: authn.IdentityTypeLoginID,
+			Claims: map[string]interface{}{
+				identity.IdentityClaimLoginIDKey:   newLoginID.Key,
+				identity.IdentityClaimLoginIDValue: newLoginID.Value,
+			},
+		},
+	}, clientID, userID)
+	if err != nil {
+		return nil, err
+	}
+
+	return f.afterAddUpdateRemoveLoginID(i)
+}
+
+func (f *WebAppFlow) afterAddUpdateRemoveLoginID(i *interaction.Interaction) (result *WebAppResult, err error) {
+	s, err := f.Interactions.GetInteractionState(i)
+	if err != nil {
+		return nil, err
+	}
+
+	// Either commit
+	if s.CurrentStep().Step == interaction.StepCommit {
+		_, err = f.Interactions.Commit(i)
+		if err != nil {
+			return
+		}
+
+		result = &WebAppResult{
+			Step: WebAppStepCompleted,
+		}
+
+		return
+	}
+
+	// Or have more steps to go through
+	step := WebAppStepSetupPassword
```

Comment:
Here we assume the next step is setting up authenticator, and the authenticator to setup is either password or OOB OTP.

It would be more robust (i.e. the interaction state would not be messed up) to assert `s.CurrentStep()` is in appropriate state (i.e. setup authenticator with correct available authenticator type).